rfcs/016-fdb-replicator.md

```diff
@@ -0,0 +1,359 @@
+
+### State Differences From CouchDB <= 3.x
+
+The set of states is slightly different than the ones from before. There are
+now less states as some of them have been combined together:
+
+ * `initializing` was combined with `pending`
+
+ * `error` was combined with `crashing`
+
+### Mapping Between couch_jobs States and Replication States
+
+`couch_jobs` application has its own set of state definitions and they map to
+replicator states like so:
+
+ | Replicator States| `couch_jobs` States
+ | ---              | :--
+ | pending          | pending
+ | running          | running
+ | crashing         | pending
+ | completed        | finished
+ | failed           | finished
+
+### State Transition Diagram
+
+Jobs start in the `pending` state, after either a `_replicator` db doc
+update, or a POST to the `/_replicate` endpoint. Continuous jobs, will
+normally toggle between `pending` and `running` states. Normal jobs
+may toggle between `pending` and running a few times and then end up
+in `completed`.
+
+```
+_replicator doc       +-------+
+POST /_replicate ---->+pending|
+                      +-------+
+                          ^
+                          |
+                          |
+                          v
+                      +---+---+      +--------+
+            +---------+running+<---->|crashing|
+            |         +---+---+      +--------+
+            |             ^
+            |             |
+            v             v
+        +------+     +---------+
+        |failed|     |completed|
+        +------+     +---------+
+```
+
+
+## Replication ID Collisions
+
+Multiple replication jobs may specify replications which map to the same
+`replication_id`. To handle these collisions there is an FDB subspace `(...,
+LayerPrefix, ?REPLICATION_IDS, replication_id) -> job_id` to keep track of
+them. After the `replication_id` is computed, each replication job checks if
+there is already another job pending or running with the same `replication_id`.
+If the other job is transient, then the current job will reschedule itself as
+`crashing`. If the other job is persistent, the current job will fail
+permanently as `failed`.
+
+## Transient Job Behavior
+
+In CouchDB <= 3.x transient replication jobs ran in memory on a particular node
+in the cluster. If the node where the replication job ran crashes, the job
+would simply disappear without a trace. It was up to the user to periodically
+monitor the job status and re-create the job. In the current design,
+`transient` jobs are persisted to FDB as `couch_jobs` records, and so would
+survive dbcore node restarts. Technically, with the new design it becomes
+possible to keep the transient job results, including completed job status to
+allow users to retrieve their state later. However in the initial
+implementation, in order to keep maximum compatibility with the <= 3.x
+replicator, transient jobs will be removed when they fail and when they
+complete.
```

Comment:
I have a hard time imagining how retaining compatibility for such an arguably anti-feature is worth it.

Reply:
It was mainly a concern about users expecting these jobs to disappear if say they are firing them up from a script.  It's also somewhat tricky that completed jobs do not appear in the list of jobs in _scheduler/jobs by default, and they do not appear in _scheduler/docs since those display only replication originating from _replicator databases.

But I could see letting these jobs linger for a configurable amount time after their completion (24 hours?) and allow querying them by their ID as _scheduler/jobs/$ID at the expense of having an extra component which scans the jobs and does that maintenance?

Reply:
Playing with running existing elixir replication tests, I notice that if we remove the jobs right away, it becomes hard to maintain the API where wait for the result of a normal replication POST-ed to _replicate.  Currently that's done using couch_job's subscription mechanism, however if the job immediately removes itself, by the time the subscription fires on the `frontend_api` node where the client is listening, the data is already gone so it's a race condition.

I think I'll add the check to the monitoring gen_server to periodically remove these transient jobs older than some configurable time period.

Reply:
Added the feature to keep failed and completed transient jobs for a configurable amount of time to the prototype branch and updated the RFC as well.